### ISSUE TITLE
--warning-mode=fail doesn't fail the build for task dependency warnings

### ISSUE THREAD

**Issue by TWiStErRob** [gradle/gradle#20843]
### Expected Behavior

`org.gradle.warning.mode=fail` fails the build, when warnings are output.
![image](https://user-images.githubusercontent.com/2906988/169889631-67536ce4-5619-4509-972c-7b28a95b23b7.png)

### Current Behavior
```
> Task :detektReportMergeSarif
Execution optimizations have been disabled for task ':detektReportMergeSarif' to ensure correctness due to the following reasons:
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\app\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':app:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\awt-hack\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:awt-hack:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\core\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:core:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\lib\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:lib:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\paparazzi\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:paparazzi:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\states\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:states:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\theme\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:theme:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\widget\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:widget:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\feature\configuration\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':feature:configuration:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
  - Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\feature\preview\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':feature:preview:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\app\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':app:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\awt-hack\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:awt-hack:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\core\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:core:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\lib\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:lib:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\paparazzi\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:paparazzi:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\states\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:states:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\theme\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:theme:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\component\widget\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':component:widget:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\feature\configuration\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':feature:configuration:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Gradle detected a problem with the following location: 'P:\projects\workspace\net.twisterrob.sun\feature\preview\build\reports\detekt\detekt.sarif'. Reason: Task ':detektReportMergeSarif' uses this output of task ':feature:preview:detekt' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed. Please refer to https://docs.gradle.org/7.4.2/userguide/validation_problems.html#implicit_dependency for more details about this problem. This behaviour has been deprecated and is scheduled to be removed in Gradle 8.0. Execution optimizations are disabled to ensure correctness. See https://docs.gradle.org/7.4.2/userguide/more_about_tasks.html#sec:up_to_date_checks for more details.
Merged SARIF output to P:\projects\workspace\net.twisterrob.sun\build\reports\detekt\merge.sarif

Execution optimizations have been disabled for 1 invalid unit(s) of work during this build to ensure correctness.
Please consult deprecation warnings for more details.

BUILD SUCCESSFUL in 14s
9 actionable tasks: 7 executed, 2 up-to-date
```

### Context
I have a clean build as of today, and want to preserve that by setting up strict checks. One of these I know about is `org.gradle.warning.mode=fail`.

### Steps to Reproduce
 * Clone https://github.com/TWiStErRob/net.twisterrob.sun
 * Check out https://github.com/TWiStErRob/net.twisterrob.sun/commit/5df93c6dee3d0449f3111c6438a26323d9df6169
 * `gradlew detekt :detektReportMergeSarif`

### Your Environment

Build scan URL: https://gradle.com/s/f5itz3aebwmvm

**Comment by wolfs** [comment]
Hi @TWiStErRob, thank you for reporting. I searched in your checkout and didn't find `org.gradle.warning.mode=fail` anywhere. Where do you set this property?

**Comment by TWiStErRob** [comment]
True... weird I think based on the title of the issue I was running with `--warning-mode=fail`, and locally (from build scan). Strange that there's no way to get exact command line from scan. I'll retry this, but I think I just missed the flag from the repro.

**Comment by TWiStErRob** [comment]
hmmmm... retried on the repro mentioned above:
| State | Command | Result | Scan URL |
| - | - | - | - |
| 5df93c6d | `./gradlew detekt :detektReportMergeSarif --scan` | success with warnings | https://gradle.com/s/jzl6cpb3ujon2 |
| 5df93c6d | `./gradlew detekt :detektReportMergeSarif --scan --warning-mode=fail` | fail | https://gradle.com/s/5zp5dsnz3dmum |
| 5df93c6d + `org.gradle.warning.mode=fail` in gradle.properties | `./gradlew detekt :detektReportMergeSarif --scan` | success with warnings | https://gradle.com/s/gub7bytefit3q |

**Comment by wolfs** [comment]
Sadly, `-P` and `gradle.properties` are not equivalent :(. I tried to reproduce it on my machine, and I was able to, though then I found the entry `org.gradle.warning.mode=all` in the `gradle.properties` file in my Gradle user home, which does override the one in the project `gradle.properties`. After removing the Gradle user home entry, the build fails as expected: https://ge.gradle.org/s/s3zpggrdpaj2i.

Do you maybe also have a `org.gradle.warning.mode` entry in your Gradle user home's `gradle.properties`?

**Comment by TWiStErRob** [comment]
Heh, your clairvoyance baffles me :) I had the exact same setup.

Based on [this list](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties) if I move my `gradle.properties` from the home to the "Gradle Installation directory", everything will be fine in the future.

Three things I would like you to follow up on, if possible:
 * Where is the "Gradle Installation Directory" when using wrappers? I searched the whole documentation and this is the only place it gets mentioned.
 * Why is a more global user-specific directory overriding a local project-specific setting?
 * Are these settings (and their sources visible in build scans?); if not, can they be?

**Comment by wolfs** [comment]
> Where is the "Gradle Installation Directory" when using wrappers? I searched the whole documentation and this is the only place it gets mentioned.

It is within the Gradle caches, so you shouldn't really change that.

> Why is a more global user-specific directory overriding a local project-specific setting?

Basically because this way you can have machine specific settings which override project specific settings. I see that for some use-cases the other way round would be better and I think there may even be an issue for that.

> Why doesn't --warning-mode=none have no effect

It does, it doesn't show the line about deprecation warnings at all in the console output. Though we still show the warning about performance degradations regardless of the deprecation warnings, since we deem them so important that we always display them. Note that the ones which show up in the console log are not deprecation warnings but a different kind of warning.

> Are these settings (and their sources visible in build scans?); if not, can they be?

They are currently not visible. Build scans always capture deprecation warnings, though it would be interesting to show them. I'll pass on the build scan feature request to the appropriate team.

I'll close the issue now, since the original problem isn't a problem.

**Comment by TWiStErRob** [comment]
Agree with the resolution, sorry for the noise and thanks for the extra info.

---

>> Where is the "Gradle Installation Directory" when using wrappers? 
>> I searched the whole documentation and this is the only place it gets mentioned.
>
> It is within the Gradle caches, so you shouldn't really change that.

I was asking because I want to put my properties file there to not have this issue again and let the project setting win. But I still don't know where it is 😞 as the caches are quite large. Is it the one where the wrapper zip gets extracted? (If we shouldn't change it, why is it kind-a documented?)

**Comment by TWiStErRob** [comment]
By the way:
>> Are these settings (and their sources visible in build scans?); if not, can they be?
> They are currently not visible. Build scans always capture deprecation warnings, though it would be interesting to show them. I'll pass on the build scan feature request to the appropriate team.

Here I meant the value that was passed in for `org.gradle.warning.mode`, and similar. And the exact command line with all the flags.

**Comment by wolfs** [comment]
The locations are documented here: https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home
If you change anything in those caches, there are no guarantees nothing bad might happen. In particular, a running daemon normally won't be happy about changes in those directories, since it assumes that no external changes will be made to the caches.

**Comment by TWiStErRob** [comment]
I'm sorry to keep coming back, but it was still not clear which folder would contain "gradle.properties in Gradle installation directory." from [this page](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_configuration_properties). The [(8) the page you linked](https://docs.gradle.org/current/userguide/directory_layout.html#dir:gradle_user_home) would be "gradle.properties in GRADLE_USER_HOME directory.".

<details><summary>So I looked at the code</summary>

*wrapper*
- `GradleWrapperMain` adds project root `gradle.properties`
- `GradleWrapperMain` adds Gradle User Home `gradle.properties`
- and that's it.

*settings*
- `ProjectPropertySettingBuildLoader` adds project root `gradle.properties`

*project*
- `DefaultGradlePropertiesLoader` adds Gradle Home `gradle.properties`
- `DefaultGradlePropertiesLoader` adds project root `gradle.properties`
- `DefaultGradlePropertiesLoader` adds Gradle User Home `gradle.properties`

Gradle Home and Gradle User Home reads from `BuildLayoutParameters`. Gradle User Home is as [documented](https://docs.gradle.org/current/userguide/build_environment.html#sec:gradle_environment_variables), and Gradle Home is calculated in `CurrentGradleInstallationLocator` with some complex logic. At this point I had the answer by attaching debugger via `-Dorg.gradle.debug=true` and I found `%GRADLE_USER_HOME%\wrapper\dists\gradle-7.4.2-all\9uukhhbclvbegdvsww0j0cr3p\gradle-7.4.2`.
</details>

and found that `%GRADLE_USER_HOME%\wrapper\dists\gradle-<version>-<type>\<hash>\gradle-\<version>` is the **Installation Directory**. The problem with this is that I can't practically put anything in there (not too worried about daemons and stuff, I can `--stop` or restart after a change if necessary). The reason for not being able is because this directory keeps changing for every project based on the gradle-wrapper.properties, so I would need to somehow hook into the wrappers that extract to make sure that my configuration properties file exist after downloading, which is infeasible.

I'm not giving up, nothing is impossible... I think I'll try to user some init scripts next, and use reflection if necessary to achieve a centralized setup that can be overridden by project/command line.

Thanks for all the info.
